server/service/src/sync/central_data_synchroniser_v6.rs

```diff
@@ -0,0 +1,97 @@
+use crate::sync::{
+    api_v6::{SyncBatchV6, SyncRecordV6},
+    sync_status::logger::SyncStepProgress,
+};
+
+use super::{
+    api::ParsingSyncRecordError,
+    api_v6::{SyncApiErrorV6, SyncApiV6},
+    sync_status::logger::{SyncLogger, SyncLoggerError},
+};
+
+use repository::{
+    CursorController, KeyValueType, RepositoryError, StorageConnection, SyncBufferRow,
+    SyncBufferRowRepository,
+};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub(crate) enum CentralPullErrorV6 {
+    #[error(transparent)]
+    SyncApiError(#[from] SyncApiErrorV6),
+    #[error("Failed to save sync buffer or cursor")]
+    SaveSyncBufferOrCursorsError(#[from] RepositoryError),
+    #[error(transparent)]
+    ParsingRecordError(#[from] ParsingSyncRecordError),
+    #[error(transparent)]
+    SyncLoggerError(#[from] SyncLoggerError),
+}
+
+#[derive(Error, Debug)]
+#[error("Failed to serialise V6 remote record into sync buffer row, record: '{record:?}'")]
+pub(crate) struct SerialisingToSyncBuffer {
+    source: serde_json::Error,
+    record: serde_json::Value,
+}
+
+pub(crate) struct CentralDataSynchroniserV6 {
+    pub(crate) sync_api_v6: SyncApiV6,
+}
+
+impl CentralDataSynchroniserV6 {
+    pub(crate) async fn pull<'a>(
+        &self,
+        connection: &StorageConnection,
+        batch_size: u32,
+        logger: &mut SyncLogger<'a>,
+    ) -> Result<(), CentralPullErrorV6> {
+        let cursor_controller = CursorController::new(KeyValueType::SyncPullCursorV6);
+        // TODO protection from infinite loop
+        loop {
+            let cursor = cursor_controller.get(&connection)?;
+
+            let SyncBatchV6 {
+                end_cursor,
+                total_records,
+                records,
+            } = self.sync_api_v6.pull(cursor, batch_size).await?;
+
+            logger.progress(SyncStepProgress::PullCentralV6, total_records)?;
+
+            let is_emtpy = records.is_empty();
+
+            for SyncRecordV6 { cursor, record } in records {
+                // Would ideally have  mapping method for this
+                let buffer_row = record.to_buffer_row()?;
+
+                insert_one_and_update_cursor(
```

Comment:
It's not important for this PR, but I feel like if catch the error instead of using `?` it should be safe to update the cursor only on  error on finishing the batch. I'm not sure if updating the cursor on every record is much overhead but feels like it could slow down the processing which I think could be a bit slow at times?
I guess it's probably more important for things like TemperatureLog Row where we might have a lot of records.

Reply:
The idea here was we update sync buffer and cursor atomically (if one fails the other fails). I added to this issue: https://github.com/msupply-foundation/open-msupply-internal/issues/39#issuecomment-1940314362